banyand/series/trace/query.go

```diff
@@ -0,0 +1,253 @@
+// Licensed to Apache Software Foundation (ASF) under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Apache Software Foundation (ASF) licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package trace
+
+import (
+	"encoding/hex"
+	"time"
+
+	flatbuffers "github.com/google/flatbuffers/go"
+	"github.com/pkg/errors"
+	"go.uber.org/multierr"
+
+	"github.com/apache/skywalking-banyandb/api/common"
+	"github.com/apache/skywalking-banyandb/api/data"
+	v1 "github.com/apache/skywalking-banyandb/api/fbs/v1"
+	"github.com/apache/skywalking-banyandb/banyand/kv"
+	"github.com/apache/skywalking-banyandb/banyand/series"
+	"github.com/apache/skywalking-banyandb/pkg/convert"
+	"github.com/apache/skywalking-banyandb/pkg/fb"
+	"github.com/apache/skywalking-banyandb/pkg/partition"
+)
+
+func (t *traceSeries) FetchTrace(traceID string, opt series.ScanOptions) (trace data.Trace, err error) {
+	if traceID == "" {
+		return trace, ErrInvalidTraceID
+	}
+	traceIDBytes := []byte(traceID)
+	traceIDShardID := partition.ShardID(traceIDBytes, t.shardNum)
+	bb, errTraceID := t.reader.TimeSeriesReader(traceIDShardID, traceIndex, 0, 0).GetAll(traceIDBytes)
+	if errTraceID != nil {
+		return trace, errTraceID
+	}
+	t.l.Debug().Uint("shard_id", traceIDShardID).
+		Str("trace_id", traceID).
+		Hex("trace_id_bytes", traceIDBytes).
+		Int("chunk_num", len(bb)).Msg("fetch Trace by trace_id")
+	if len(bb) < 1 {
+		return trace, nil
+	}
+	chunkIDs := make([]common.ChunkID, len(bb))
+	for i, b := range bb {
+		chunkIDs[i] = b
+	}
+	entities, errEntity := t.FetchEntity(chunkIDs, opt)
+	if errEntity != nil {
+		return trace, errEntity
+	}
+	return data.Trace{
+		KindVersion: data.TraceKindVersion,
+		Entities:    entities,
+	}, err
+}
+
+func (t *traceSeries) ScanEntity(startTime, endTime uint64, opt series.ScanOptions) ([]data.Entity, error) {
+	total := opt.Limit
+	if total < 1 {
+		total = 10
+	}
+	states := make([]byte, 0, 2)
+	switch opt.State {
+	case series.TraceStateSuccess:
+		states = append(states, StateSuccess)
+	case series.TraceStateError:
+		states = append(states, StateError)
+	default:
```

Comment:
I guess I've added a default `TraceState`. Do I need to remove this default field?

Reply:
done